File: worker/python/dispatcher.py
"""
isolane/worker/python/dispatcher.py

Routes work items from Redis Streams to the engine.

The dispatcher is the bridge between the Redis consumer and the
batch processing engine. It:
  1. Deserialises the work item payload
  2. Fetches a DB connection scoped to the tenant namespace
  3. Calls engine/batch.py process_batch()
  4. ACKs the message on success
  5. Does NOT ACK on failure — XAUTOCLAIM will retry

Work item fields (set by the producer / seed script):
  pipeline_id       — which pipeline to run
  batch_offset      — Kafka offset or batch reference
  records_json      — JSON-encoded list of record dicts
  _claim_count      — incremented on each XAUTOCLAIM re-delivery
"""

import asyncio
import json
import os
from typing import Any, Optional

import asyncpg

from engine.batch import process_batch, BatchResult
from engine.rocksdb_client import get_client
from worker.python.metrics import (
    batch_processed_total,
    quarantine_total,
    batch_duration_seconds,
    touch_namespace,
)


async def _get_db_conn(namespace: str) -> asyncpg.Connection:
    """
    Get a DB connection with tenant context set.
    """
    conn = await asyncpg.connect(
        host     = os.environ.get("POSTGRES_HOST", "postgres"),
        port     = int(os.environ.get("POSTGRES_PORT", 5432)),
        user     = os.environ.get("POSTGRES_USER", "mtpif"),
        password = os.environ.get("POSTGRES_PASSWORD", "mtpif_dev"),
        database = os.environ.get("POSTGRES_DB", "mtpif"),
    )
    await conn.execute("SELECT set_tenant_context($1)", namespace)
    return conn


async def dispatch_async(
    namespace:  str,
    stream_id:  str,
    fields:     dict[str, str],
) -> Optional[BatchResult]:
    """
    Async dispatch — called by the consumer for each work item.
    Returns BatchResult on success, None on failure.
    """
    touch_namespace(namespace)

    pipeline_id  = fields.get("pipeline_id", "unknown")
    batch_offset = fields.get("batch_offset", stream_id)

    # Deserialise records from JSON
    records_raw = fields.get("records_json", "[]")
    try:
        records: list[dict[str, Any]] = json.loads(records_raw)
    except json.JSONDecodeError as e:
        print(
            f"[dispatcher] {namespace}.{pipeline_id} | "
            f"invalid records_json: {e}"
        )
        return None

    if not records:
        print(
            f"[dispatcher] {namespace}.{pipeline_id} | "
            f"empty batch — skipping"
        )
        return None

    rdb  = get_client()
    conn = await _get_db_conn(namespace)

    try:
        result = await process_batch(
            conn              = conn,
            rdb               = rdb,
            namespace         = namespace,
            pipeline_id       = pipeline_id,
            records           = records,
            batch_offset      = batch_offset,
            stream_message_id = stream_id,
        )

        # Update Prometheus metrics
        batch_processed_total.labels(pipeline=pipeline_id).inc(
            result.records_out
        )
        if result.records_quarantined > 0:
            quarantine_total.labels(pipeline=pipeline_id).inc(
                result.records_quarantined
            )
        batch_duration_seconds.labels(pipeline=pipeline_id).observe(
            result.duration_ms / 1000
        )

        return result

    except Exception as e:
        print(
            f"[dispatcher] {namespace}.{pipeline_id} | "
            f"unhandled error: {e}"
        )
        return None
    finally:
        await conn.close()


def dispatch(
    namespace: str,
    stream_id: str,
    fields:    dict[str, str],
) -> Optional[BatchResult]:
    """
    Sync wrapper around dispatch_async.
    """
    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            # Already in an async context — create a task
            import concurrent.futures
            with concurrent.futures.ThreadPoolExecutor() as pool:
                future = pool.submit(
                    asyncio.run,
                    dispatch_async(namespace, stream_id, fields)
                )
                return future.result(timeout=300)
        else:
            return loop.run_until_complete(
                dispatch_async(namespace, stream_id, fields)
            )
    except Exception as e:
        print(f"[dispatcher] {namespace} | dispatch error: {e}")
        return None

File: worker/go/dispatcher.go
// isolane/worker/go/dispatcher.go
//
// Routes work items to the Python engine subprocess.
//
// The Go worker owns the Redis consumer loop, XAUTOCLAIM,
// and ACK logic. The Python engine owns schema validation,
// edge case checks, quarantine writes, staging writes, and dbt.
//

package worker

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// EngineResult is the JSON response from the Python engine subprocess.
type EngineResult struct {
	Status             string  `json:"status"`
	RecordsIn          int     `json:"records_in"`
	RecordsOut         int     `json:"records_out"`
	RecordsQuarantined int     `json:"records_quarantined"`
	DurationMs         float64 `json:"duration_ms"`
	ErrorMessage       string  `json:"error_message"`
}

// Dispatch calls the Python engine for a single work item.
// Returns a DispatchResult summarising the outcome.
func Dispatch(
	ctx context.Context,
	item *WorkItem,
	cfg *Config,
) *DispatchResult {
	start := time.Now()

	result := &DispatchResult{
		WorkItem: item,
	}

	// Call Python engine subprocess
	engineResult, err := callPythonEngine(ctx, item, cfg)
	if err != nil {
		result.Status = "failed"
		result.Error = err
		result.Duration = time.Since(start)
		fmt.Printf(
			"[dispatcher] %s.%s | engine error: %v\n",
			item.Namespace, item.PipelineID(), err,
		)
		return result
	}

	result.Status = engineResult.Status
	result.RecordsIn = engineResult.RecordsIn
	result.RecordsOut = engineResult.RecordsOut
	result.RecordsQuarantined = engineResult.RecordsQuarantined
	result.Duration = time.Since(start)

	if engineResult.ErrorMessage != "" {
		result.Error = fmt.Errorf("%s", engineResult.ErrorMessage)
	}

	// Update Prometheus metrics
	if result.RecordsOut > 0 {
		BatchProcessedTotal.WithLabelValues(item.PipelineID()).
			Add(float64(result.RecordsOut))
	}
	if result.RecordsQuarantined > 0 {
		QuarantineTotal.WithLabelValues(item.PipelineID()).
			Add(float64(result.RecordsQuarantined))
	}
	BatchDurationSeconds.WithLabelValues(item.PipelineID()).
		Observe(result.Duration.Seconds())

	fmt.Printf(
		"[dispatcher] %s.%s | %s | in=%d out=%d quarantined=%d duration=%dms\n",
		item.Namespace, item.PipelineID(), result.Status,
		result.RecordsIn, result.RecordsOut, result.RecordsQuarantined,
		result.Duration.Milliseconds(),
	)

	return result
}

// callPythonEngine invokes the Python engine as a subprocess.
func callPythonEngine(
	ctx context.Context,
	item *WorkItem,
	cfg *Config,
) (*EngineResult, error) {
	args := []string{
		"-m", "engine.dispatch",
		"--namespace", item.Namespace,
		"--pipeline", item.PipelineID(),
		"--stream-id", item.StreamID,
		"--batch-offset", item.BatchOffset(),
	}

	cmd := exec.CommandContext(ctx, "python3", args...)

	// Pass records via stdin to avoid shell argument length limits
	cmd.Stdin = strings.NewReader(item.RecordsJSON())

	// Set environment
	cmd.Env = buildEnv(cfg)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		errMsg := stderr.String()
		stdoutMsg := stdout.String()
		if len(errMsg) > 1000 {
			errMsg = errMsg[len(errMsg)-1000:]
		}
		return nil, fmt.Errorf(
			"python engine exited %v\nSTDERR: %s\nSTDOUT: %s",
			err, errMsg, stdoutMsg,
		)
	}

	// Extract the last non-empty line — that's the JSON result
	// Previous lines are print() output from the engine (logs)
	outputLines := strings.Split(strings.TrimSpace(stdout.String()), "\n")
	var jsonLine string
	for i := len(outputLines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(outputLines[i])
		if len(line) > 0 && line[0] == '{' {
			jsonLine = line
			break
		}
	}
	if jsonLine == "" {
		return nil, fmt.Errorf(
			"no JSON found in engine output: %s",
			stdout.String()[:min(500, len(stdout.String()))],
		)
	}

	var result EngineResult
	if err := json.Unmarshal([]byte(jsonLine), &result); err != nil {
		return nil, fmt.Errorf(
			"failed to parse engine output: %v (line: %s)",
			err, jsonLine,
		)
	}

	return &result, nil
}

func buildEnv(cfg *Config) []string {
	// Inherit the full process environment so PATH, HOME, etc. are available
	// then override with our specific values
	env := os.Environ()
	overrides := map[string]string{
		"PYTHONPATH":           "/app",
		"PYTHONUNBUFFERED":     "1",
		"POSTGRES_HOST":        cfg.PostgresHost,
		"POSTGRES_PORT":        fmt.Sprintf("%d", cfg.PostgresPort),
		"POSTGRES_USER":        cfg.PostgresUser,
		"POSTGRES_PASSWORD":    cfg.PostgresPass,
		"POSTGRES_DB":          cfg.PostgresDB,
		"REDIS_HOST":           cfg.RedisHost,
		"REDIS_PORT":           fmt.Sprintf("%d", cfg.RedisPort),
		"DBT_PROJECT_DIR":      cfg.DBTProjectDir,
		"ROCKSDB_PATH":         cfg.RocksDBPath,
		"TENANT_ROLE_PASSWORD": cfg.TenantRolePass,
	}

	// Apply overrides
	result := make([]string, 0, len(env)+len(overrides))
	seen := make(map[string]bool)
	for k := range overrides {
		seen[k] = true
	}
	for _, e := range env {
		key := e[:indexOf(e, '=')]
		if !seen[key] {
			result = append(result, e)
		}
	}
	for k, v := range overrides {
		result = append(result, fmt.Sprintf("%s=%s", k, v))
	}
	return result
}

func indexOf(s string, c byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			return i
		}
	}
	return len(s)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// ParseInt safely parses a string to int with a default.
func ParseInt(s string, def int) int {
	n, err := strconv.Atoi(s)
	if err != nil {
		return def
	}
	return n
}
